src/main/java/org/cactoos/iterator/Mapped.java

```diff
@@ -24,53 +24,59 @@
 package org.cactoos.iterator;
 
 import java.util.Iterator;
+import java.util.NoSuchElementException;
 import org.cactoos.Func;
-import org.cactoos.func.UncheckedFunc;
+import org.cactoos.scalar.Unchecked;
 
 /**
  * Mapped iterator.
  *
- * <p>There is no thread-safety guarantee.
+ * <p>
+ * There is no thread-safety guarantee.
  *
- * @param <X> Type of source item
  * @param <Y> Type of target item
  * @since 0.1
  */
-public final class Mapped<X, Y> implements Iterator<Y> {
-
-    /**
-     * Iterator.
-     */
-    private final Iterator<X> origin;
-
-    /**
-     * Function.
-     */
-    private final Func<X, Y> fnc;
+public final class Mapped<Y> extends IteratorEnvelope<Y> {
 
     /**
      * Ctor.
      * @param func Func
      * @param iterator Source iterator
+     * @param <X> Type of item
+     * @checkstyle AnonInnerLengthCheck (60 lines)
      */
-    public Mapped(final Func<X, Y> func, final Iterator<X> iterator) {
-        this.origin = iterator;
-        this.fnc = func;
-    }
+    public <X> Mapped(
+        final Func<? super X, ? extends Y> func,
+        final Iterator<? extends X> iterator
+    ) {
+        super(
+            new Iterator<Y>() {
 
-    @Override
-    public boolean hasNext() {
-        return this.origin.hasNext();
-    }
+                @Override
+                public boolean hasNext() {
+                    return iterator.hasNext();
+                }
 
-    @Override
-    public Y next() {
-        return new UncheckedFunc<>(this.fnc).apply(this.origin.next());
-    }
+                @Override
+                public Y next() {
+                    if (this.hasNext()) {
+                        return new Unchecked<>(
+                            new org.cactoos.scalar.Mapped<>(
+                                func, iterator::next
+                            )
+                        ).value();
+                    }
+                    throw new NoSuchElementException();
```

Comment:
@andreoss why change the implementation of `next()`? Wasn't the previous one correct?

Reply:
@victornoel I wanted to reuse `scalar.Mapped` here, to be consistent between different Mapped implementations. 
FindBugs was triggered by the lack of a `hasNext` check, not sure why it used to accept  the old code

Reply:
@andreoss I see, personally I find that `scalar.Mapped` makes things harder to read but it is correct anyway, let's keep it like you wrote it :)